### PR TITLE
Next iteration including new endpoints and validation

### PR DESCRIPTION
**Pull request by mscappini**

### CODE REVIEW

File: lib/app.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 (function() {
   'use strict';
 
-  process.env.NODE_ENV = process.env.NODE_ENV || 'dev';
+  process.env.NODE_ENV = process.env.NODE_ENV || 'development';
```
Comment (aisthesis) [inline lib/app.js]: Are you sure you need this? I noticed when I changed the value to 'production' on my VMI, all of a sudden dev dependencies weren't pulled.  `npm` will default to 'development', so if you're not actually reading it, I don't think you need it.

File: package.json [hunk 2]
```diff
@@ -1,39 +1,40 @@
...
+  "config": {
+    "host": "",
+    "port": "8081",
+    "db": {
+      "optionsMkt": "mongodb://localhost:27017/optionsMkt"
     }
+  },
+  "dependencies": {
+    "express": "~4.12.3",
+    "moment": "^2.10.2",
```
Comment (aisthesis) [inline package.json]: use tilde

**Comment by aisthesis** [comment]
Include in the README under testing the code conformity tests, as in https://github.com/optbot/template#code-conformity

**Comment by aisthesis** [comment]
When I try to start it, I'm getting:

```
$ npm start

> @optbot/restapi@0.0.2 start /mnt/disk1/home/marshall/workspace/restapi
> node lib/app.js

[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
Express server started at http://:::8081
```

And since the process is blocking and seems to be running despite the errors, I tried hitting it through the browser with `http://192.168.56.101:8081/api/?equity=SCTY&option=call&strike=30&expiry=2015-4-9`. But I'm getting the simple 'Not found' page (unformatted, so it's picking up that from the `restapi` code).

**Comment by mscappini** [comment]
Oops, forgot to update the readme. I added new endpoints to be more specific. Try `/api/options/?...` to test.

Those warnings are related to the Mongo install not finding python. I've only ever had the issue on Windows, never seen it on Mac or Ubuntu.

**Comment by aisthesis** [comment]
now i get `{"options":[]}`

**Comment by aisthesis** [comment]
Those _errors_ really need to be gone before we check it in, too. We can't release code known to generate errors, and warnings should only happen under special circumstances, certainly not under `npm start`

File: lib/routes/api.js [hunk 3]
```diff
@@ -1,80 +1,129 @@
 var querystring = require('querystring');
+var moment = require('moment');
 var express = require('express');
 var router = express.Router();
 
 var MongoClient = require('mongodb').MongoClient;
```
Comment (aisthesis) [inline lib/routes/api.js]: The instance of the MongoClient class should be camel case.

**Comment by mscappini** [comment]
It's actually Mongo that gives the errors. Like I said, I've never seen that happen on anything other than Windows so I just assumed it was a Windows issue. Apparently not.

Plus the app will still run with the errors. That's why I consider them warnings. But I guess they are technically still errors...

File: lib/app.js (hunk 1 above)
Comment (mscappini) [inline lib/app.js]: Well, by default it's undefined. Under scenarios where Node.js sees that it's undefined, it assumes development. I suppose our code could work the same way, but I was putting development here to be more specific.

**Comment by mscappini** [comment]
Regarding the empty options, is your database still called `options`? It's looking for a database called `optionsMkt` but will return no data if the database is named incorrectly. This is how Mongo works out of the box, we would have to specifically check for the database and then throw an exception if it doesn't exist if that's the behavior we're looking for.

**Comment by aisthesis** [comment]
yes, and it will continue to be called options for a bit at least. Could you call out in the README how to configure it appropriately?

**Comment by aisthesis** [comment]
Errors are gone after re-installing the node_modules folder.

**Comment by aisthesis** [comment]
One other thing: Could you make the 'start' instruction in the README be `sudo npm start`? I realize that it doesn't matter at the moment for this service, but `sudo` should at least not break it.

The reason I'd like it as that each service that stays up should run as its own user and group, so if you or I start it as ourselves rather than root, we actually shouldn't be able to. Root will have the ability to act on behalf of any user and group, so also the one running the service. I've implemented this for `eqpuller` following the way `mongo` sets up its running process using the user and group `mongodb`.

Also the `npm config` commands given as examples won't work without `sudo` because global configurations are stored under root control. So, could you add `sudo` to those example commands as well.

File: README.md [hunk 4]
```diff
@@ -33,11 +33,22 @@ Testing
...
 If the service is running on port `8081` on a host with IP address `91.198.174.192`, navigate to the following REST endpoint in a web browser:
 
-	http://91.198.174.192:8081/api/?equity=SCTY&option=call&strike=30&expiry=2015-4-9
+	http://91.198.174.192:8081/api/options?equity=SCTY&option=call&strike=30&expiry=2015-4-9
 
 You can replace query string parameters as needed to query for suitable data.
 
+If no data is being returned, it may be that the database connection string is improperly configured. As outlined in the **Connects To** section, this project currently requires a database called `optionsMkt` with a collection named `quotes`. To reconfigure this, you can issue the following command:
+
+	$ npm config set @optbot/restapi:db:optionsMkt "mongodb://<your_mongo_host>:<your_mongo_port>/<your_database_name>"
```
Comment (aisthesis) [inline README.md]: missing `sudo` and `--global`

File: README.md [hunk 5]
```diff
@@ -33,11 +33,22 @@ Testing
...
+
+As an example, using the default settings would be
+
+	$ npm config set @optbot/restapi:db:optionsMkt "mongodb://localhost:27017/optionsMkt"
```
Comment (aisthesis) [inline README.md]: missing `sudo` and `--global`

**Comment by mscappini** [comment]
I would prefer to not have `sudo` in the instructions because I am trying to keep the instructions as platform-agnostic as possible. So far, these instructions will work on Windows as well. I'm sure if you're using some form of Linux, you know what `sudo` does and will use it where appropriate.

In a related note: I was reading that `sudo` is [not recommended](https://github.com/npm/npm/tree/99f804f43327c49ce045ae2c105995636c847145#permissions) for `npm start`.

**Comment by aisthesis** [comment]
How will we be able to run the process as user `restapi` and group `restapi` otherwise?

**Comment by aisthesis** [comment]
I ran `sudo npm config set @optbot/restapi:db:optionsMkt "mongodb://localhost:27017/options" --global` but am still getting no data.

**Comment by aisthesis** [comment]
The reason I prefer `sudo` in the README is because, while it is ok to be messing around with instructions in development, installing and running on production should be pure copy-paste. And the production server is Linux. When I install there, I want to know exactly what to do without having to think about it. On Windows it is easily recognizable that sudo won't work, you can leave it out on your copy-paste, and it simply won't do anything if you forget. On Linux, it is possible that leaving it out will do a bunch of stuff that you don't want to the user that is running `npm start` because the system _will_ understand what it means to run it without `sudo`.

I wasn't able to find anything about not running `npm start` as root in the link you gave. All I saw was that it is ok to run some commands as root.

**Comment by mscappini** [comment]
But you could make the same argument for *nix. You should know to run as root, so include `sudo`. Basically, do what is appropriate for your OS. I don't believe README instructions should be a copy/paste job. They are guidelines, and things should work out of the box, but setting up without even thinking just seems like a mistake to begin with.

**Comment by mscappini** [comment]
But I also have another thought on that. It's been to my understanding that you should never run a webserver as root for security reasons. Wouldn't using `sudo` run this webserver as root?

**Comment by aisthesis** [comment]
The instructions should apply to the production servers and should be formed in such a way to later be automated. The reason for including `sudo` now is to make sure there are no unanticipated side effects that don't show actual errors.

If, on production, I happen to type `npm config set foo bar` and then `npm start` with no `sudo`, then the process will run for me. But then the configuration won't apply if you go to the same server because the config settings will be saved in files specific to my user.  Then perhaps you'll go and do your config in the same way, and soon we have the same service on the same server running in different ways depending on who starts it. The problem on Linux is that it's non-trivial to know when to use it and when not (in contrast to the 'nix' command, where the answer is always 'not').

**Comment by aisthesis** [comment]
As to your point of running it as root, the point is actually to avoid that. The difficulties I had with `eqpuller` involved creating a specific `eqpuller` user and a specific `eqpuller` group that can write to the `eqpuller` logs and run the `eqpuller` process but can do nothing else. That's how we should configure these systems in my opinion. For now, if I don't use sudo, I'll be starting it on production as 'marshall', which is also wrong. What `sudo` does is allow you freedom to specify what user is running the process when you start it. For example: https://github.com/optbot/eqpuller/blob/marshall/lib/start.js#L24

**Comment by mscappini** [comment]
I understand the desire to want `sudo` for that reason. But when I setup an IIS server, I always run it with an app pool using a particular user. This is knowledge that I know and understand as someone who sets up websites on Windows. The same knowledge should still apply to you as someone who sets up a server running on any OS you desire to run this instance on--including Linux. In which case, if you are doing this, you would know to use `sudo` where appropriate.

**Comment by mscappini** [comment]
I understand using `sudo` to specify the user. You should do that, and also know to use `sudo` to run it as that user.

**Comment by mscappini** [comment]
The point about automation I believe to actually be in favor of not putting `sudo` in the instructions. If it's automated, all you'll have to do is remember to include it once. Obviously this will be tested thoroughly before we continue to automate everything, so having to write it into the script only once shouldn't be a hassle.

**Comment by mscappini** [comment]
And wouldn't using `sudo` to run as a user negate the copy/paste anyway? You have to specify the user to run as.

**Comment by aisthesis** [comment]
No, the creation of the user and the user with which to run the code all happen in code. [This](https://github.com/mongodb/mongo/blob/7423c26ab95464a9b8f620ab4a11c297dc0621d5/debian/mongodb-org-server.postinst#L24) is how mongo accomplishes the user creation, for example, which is why you have to run the installation as `sudo`. We don't have anything nearly as elaborate as mongo, but to do it right, we do have to create a new user with permissions to run the appropriate executables and with permissions to write to the appropriate log directory.

**Comment by mscappini** [comment]
I see, can you point me to where this all takes place so I can better understand it?

**Comment by aisthesis** [comment]
https://github.com/optbot/eqpuller/blob/marshall/lib/setup.js

I wouldn't worry about it too much for now. I know you prefer developing on Windows, and this is a Linux thing. I'm happy to check out the code and make the necessary changes when the time comes. For now, since it's only running temporarily, it's no big deal anyway, and I'm not worried about who owns the process.

Once the service is publicly exposed and it does become an issue, It might be possible to set a flag on `npm start` that bypasses platform-specific code so you can continue to develop on Windows if you like. I'd be for that as long as the branching stays fairly tame. The commands for user creation and managing processes under a certain user are clearly not going to work on both platforms, but we could have something like `npm start --win` that runs it under Windows, bypassing the user-creation steps and similar.

Again, this isn't meant for this iteration but for a few cycles down the road. For this iteration, I just want to know what I'm doing when I install in production without having a separate text file of my own to record the commands.

**Comment by mscappini** [comment]
Sounds good. I'll have a look when I get back to a computer.

I fixed the query returning no results. As usual, it was a timezone thing.

**Comment by aisthesis** [comment]
It's working now. My db on VirtualBox just isn't populated with the expiry date of the example.

**Comment by mscappini** [comment]
Excellent. I wish I had an example that would always certainly work.

**Comment by aisthesis** [comment]
Ship it!

**Pull request merged**